### PR TITLE
Transition all endpoint tests in agent_endpoint_test.go to go through ServeHTTP

### PR DESCRIPTION
**Pull request by Mathew-Estafanous**
Continues issue #11396 

**Overview:**
The goal of this PR is to transition the rest of the endpoint tests in ``agent_endpoint_test.go`` to use ServeHTTP. This is going to be a large change so I’m hoping to get some progressive feedback.

### CODE REVIEW

**Comment by Mathew-Estafanous** [comment]
@kisunji I’m ready for a code review for this PR. However there are two issues that I’d like to bring up. First is the failing ``HealthServiceByID`` test case that I mention earlier in this PR. I’m not sure how to go about fixing it.

Secondly, the handler for ``AgentHost`` does not write to the response writer and as a result I cannot transition those tests to use ServeHTTP.

**Comment by Mathew-Estafanous** [comment]
I fixed the failing ``HealthServiceByID`` test case since the response’s ``CodeWithPayloadError`` was setting the ContentType to application/json when the response was suppose to be text/plain.

**Review by eculver — COMMENTED** [review]
This looks good, especially since it even uncovered a bug in the response headers! We should let @kisunji have another pass at it though since he's been in this code recently and may have some opinions.

**Comment by Mathew-Estafanous** [comment]
@eculver sorry to ping you, but I’m wondering when I should expect to have this PR reviewed and merged? It’s been sometime and I haven’t received much news regarding this PR.

**Comment by kisunji** [comment]
> @eculver sorry to ping you, but I’m wondering when I should expect to have this PR reviewed and merged? It’s been sometime and I haven’t received much news regarding this PR.

Hey @Mathew-Estafanous, I've been away on break but I'm back and will take a look this week!

**Comment by kisunji** [comment]
This is quite the comprehensive change! Thanks a lot for this, including the bugfix and the small cleanups in the tests.

I won't get through everything this week but will continue to review next week!

File: agent/agent_endpoint_test.go [hunk 1]
```diff
@@ -6787,6 +6828,7 @@ func TestAgent_Host(t *testing.T) {
 	testrpc.WaitForLeader(t, a.RPC, "dc1")
 	req, _ := http.NewRequest("GET", "/v1/agent/host?token=master", nil)
 	resp := httptest.NewRecorder()
+	// TODO: AgentHost should write to response so that we can test using ServeHTTP()
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:6831]: Thanks for the TODO! Definitely a valuable one here.

**Review by kisunji — CHANGES_REQUESTED** [review]
Thanks for your patience @Mathew-Estafanous! Went through it and found a few minor changes but the rest looks good 👍

**Comment by kisunji** [comment]
Ah there seems to be some merge conflicts 😢 Try rebasing from main and preserve this line that was added in https://github.com/hashicorp/consul/pull/11610
```
require.Empty(resp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
```

**Comment by Mathew-Estafanous** [comment]
Thanks for the review @kisunji! I know this was a big change to go through.

So I first rebased and did my best to solve each merge conflict. I think everything went well, however there are a few new test cases that will now need to be changed to use ServeHTTP. I’ll try to get them done in the next hour or two.

Regarding ``int(val.DebugConfig["SerfPortLAN"].(float64))``, I believe I had to do that because json.Unmarshal seems to unmarshal numbers as float64. As a result I have to typecast to float64 before converting to int.

I _think(?)_ I fixed the issue with the two test cases that do not properly handle non-200 responses. If I didn’t, could you maybe elaborate a little more on what you meant?

**Comment by Mathew-Estafanous** [comment]
I think I just discovered that I have a failing test case and I think it happened when I was “fixing” the merge conflicts. 😞

**Comment by kisunji** [comment]
Nice! You got what I meant 👍 

Regarding the failing tests, https://github.com/hashicorp/consul/pull/11610 actually changed some of the assertions around cache HIT or MISS so you may have to redo the rebase or compare assertions with the PR.

Feel free to tag me for any questions or help!

**Comment by Mathew-Estafanous** [comment]
@kisunji 
I was looking at the main branch and noticed t hat ``require.Equal("HIT", resp.Header().Get("X-Cache”))`` is only present in the ``TestAgentConnectAuthorize_allow`` tests. However, in my branch I somehow have them in many other tests, so I assume I’m supposed to remove all the other instances and that seems to fix the failing tests.

I’m I approaching this correctly?

File: agent/agent_endpoint_test.go [hunk 2]
```diff
@@ -357,17 +375,20 @@ func TestAgent_Services_ACLFilter(t *testing.T) {
...
+			t.Fatalf("Err: %v", err)
+		}
 
-		val := obj.(map[string]*api.AgentService)
-		require.Empty(val)
-		require.Empty(rsp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
+		if len(val) != 0 {
+			t.Fatalf("bad: %v", val)
+		}
+		require.Empty(t, resp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:391]: Try pushing a commit reverting the changes for `TestAgent_Services_ACLFilter` (I believe these tests were added from the rebase) and see if the tests pass

Comment (kisunji) [reply agent/agent_endpoint_test.go:391]: It looks like these following tests were added in the rebase:
```
TestAgentConnectCALeafCert
TestAgent_Members
TestAgent_Checks
TestAgent_Services
```
Could you try restoring them to `main` first?

Comment (kisunji) [reply agent/agent_endpoint_test.go:391]: Actually, ignore that! I think I see the issue. Will comment on the PR

File: agent/agent_endpoint_test.go [hunk 3]
```diff
@@ -391,17 +412,19 @@ func TestAgent_Services_ACLFilter(t *testing.T) {
...
+		var val map[string]*api.AgentService
+		err := dec.Decode(&val)
+		if err != nil {
+			t.Fatalf("Err: %v", err)
+		}
 
-		val := obj.(map[string]*api.AgentService)
-		require.Len(val, 2)
-		require.Empty(rsp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
+		if len(val) != 1 {
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:425]: The previous assertion was for len=2, not 1

Also, I'd prefer using package `require` here for readability

File: agent/agent_endpoint_test.go [hunk 4]
```diff
@@ -1378,17 +1414,18 @@ func TestAgent_Checks_ACLFilter(t *testing.T) {
...
 
-		val := obj.(map[types.CheckID]*structs.HealthCheck)
-		require.Len(val, 2)
-		require.Empty(rsp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
+		dec := json.NewDecoder(resp.Body)
+		val := make(map[types.CheckID]*structs.HealthCheck)
+		if err := dec.Decode(&val); err != nil {
+			t.Fatalf("Err: %v", err)
+		}
+		if len(val) != 1 {
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:1426]: Same issue here, we want len=2

File: agent/agent_endpoint_test.go [hunk 5]
```diff
@@ -1979,17 +2030,18 @@ func TestAgent_Members_ACLFilter(t *testing.T) {
...
 
-		val := obj.([]serf.Member)
-		require.Len(val, 2)
-		require.Empty(rsp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
+		dec := json.NewDecoder(resp.Body)
+		val := make([]serf.Member, 0)
+		if err := dec.Decode(&val); err != nil {
+			t.Fatalf("Err: %v", err)
+		}
+		if len(val) != 1 {
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:2042]: We want len=2

**Review by kisunji — COMMENTED** [review]
Not sure what is going on with 
```
TestAgentConnectCALeafCert_good
TestAgentConnectCALeafCert_goodNotLocal
TestAgentConnectCALeafCert_secondaryDC_good
```
But maybe try reverting the changes in those tests to see if they pass 🤔

File: agent/agent_endpoint_test.go [hunk 6]
```diff
@@ -6109,9 +6139,14 @@ func TestAgentConnectCALeafCert_goodNotLocal(t *testing.T) {
...
-		require.NoError(err)
-		require.Equal(obj, obj2)
+		a.srv.h.ServeHTTP(resp, req)
+		dec := json.NewDecoder(resp.Body)
+		issued2 := &structs.IssuedCert{}
+		require.NoError(dec.Decode(issued2))
+		require.Equal(issued, issued2)
+
+		// Should cache hit this time and not make request
+		require.Equal("HIT", resp.Header().Get("X-Cache"))
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:6149]: Let's delete this

File: agent/agent_endpoint_test.go [hunk 7]
```diff
@@ -5968,9 +5993,14 @@ func TestAgentConnectCALeafCert_good(t *testing.T) {
...
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:6003]: Let's delete this

File: agent/agent_endpoint_test.go [hunk 8]
```diff
@@ -6383,9 +6428,14 @@ func TestAgentConnectCALeafCert_secondaryDC_good(t *testing.T) {
...
-		require.NoError(err)
-		require.Equal(obj, obj2)
+		a2.srv.h.ServeHTTP(resp, req)
+		dec := json.NewDecoder(resp.Body)
+		issued2 := &structs.IssuedCert{}
+		require.NoError(dec.Decode(issued2))
+		require.Equal(issued, issued2)
+
+		// Should cache hit this time and not make request
+		require.Equal("HIT", resp.Header().Get("X-Cache"))
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:6438]: Let's delete this

**Review by kisunji — COMMENTED** [review]
Let's leave out the cache assertions if they didn't exist before

**Comment by Mathew-Estafanous** [comment]
> Let's leave out the cache assertions if they didn't exist before

Yup I’ll do that. Looks like those cache assertions are the failing points of almost all of the failing test cases.

File: agent/agent_endpoint_test.go [hunk 9]
```diff
@@ -1968,28 +2017,29 @@ func TestAgent_Members_ACLFilter(t *testing.T) {
...
 
-		val := obj.([]serf.Member)
-		require.Len(val, 2)
-		require.Empty(rsp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
+		dec := json.NewDecoder(resp.Body)
+		val := make([]serf.Member, 0)
+		if err := dec.Decode(&val); err != nil {
+			t.Fatalf("Err: %v", err)
+		}
+		require.Len(t, val, 1)
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:2042]:
```suggestion
		require.Len(t, val, 2)
```

File: agent/agent_endpoint_test.go [hunk 10]
```diff
@@ -1367,28 +1403,29 @@ func TestAgent_Checks_ACLFilter(t *testing.T) {
...
 
-		val := obj.(map[types.CheckID]*structs.HealthCheck)
-		require.Len(val, 2)
-		require.Empty(rsp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
+		dec := json.NewDecoder(resp.Body)
+		val := make(map[types.CheckID]*structs.HealthCheck)
+		if err := dec.Decode(&val); err != nil {
+			t.Fatalf("Err: %v", err)
+		}
+		require.Len(t, val, 1)
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:1428]:
```suggestion
		require.Len(t, val, 2)
		require.Empty(resp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
```

File: agent/agent_endpoint_test.go [hunk 11]
```diff
@@ -380,28 +401,32 @@ func TestAgent_Services_ACLFilter(t *testing.T) {
...
+		dec := json.NewDecoder(resp.Body)
+		var val map[string]*api.AgentService
+		err := dec.Decode(&val)
+		if err != nil {
+			t.Fatalf("Err: %v", err)
+		}
 
-		val := obj.(map[string]*api.AgentService)
-		require.Len(val, 2)
-		require.Empty(rsp.Header().Get("X-Consul-Results-Filtered-By-ACLs"))
```
Comment (kisunji) [inline agent/agent_endpoint_test.go:404]: Should preserve these lines

**Review by kisunji — APPROVED** [review]
I think these should be the only leftover TODOs! Thanks for your patience

**Comment by Mathew-Estafanous** [comment]
@kisunji sorry I just pushed another commit 😅 I forgot to add the ``require.NotEmpty(t, resp.Header().Get("X-Consul-Results-Filtered-By-ACLs”))`` line for each root-token test.

**Comment by kisunji** [comment]
> @kisunji sorry I just pushed another commit 😅 I forgot to add the `require.NotEmpty(t, resp.Header().Get("X-Consul-Results-Filtered-By-ACLs”))` line for each root-token test.

I think it should be `require.Empty` 😛

**Comment by Mathew-Estafanous** [comment]
@kisunji ohh boy....I think I’m getting a little tooo excited to push changes 😅 I’ll fix that. Sooo sorry

**Review by kisunji — APPROVED** [review]
😆 Happens to me all the time.
Looks great! We're currently in a code freeze for a release but I'll merge your PR shortly afterwards

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/519921.

**Pull request merged**